File: Software/dashboard/packages/dashboard_host/lib/src/protocol/cbor.dart
import 'dart:convert';
import 'dart:typed_data';

class ProtocolCodecError implements Exception {
  ProtocolCodecError(this.message);
  final String message;
  @override
  String toString() => 'ProtocolCodecError: $message';
}

const _majorUint = 0;
const _majorNegint = 1;
const _majorBytes = 2;
const _majorText = 3;
const _majorArray = 4;
const _majorMap = 5;
const _majorSimple = 7;

Uint8List _concat(List<Uint8List> chunks) {
  final total = chunks.fold<int>(0, (n, c) => n + c.length);
  final out = Uint8List(total);
  var offset = 0;
  for (final chunk in chunks) {
    out.setRange(offset, offset + chunk.length, chunk);
    offset += chunk.length;
  }
  return out;
}

Uint8List encodeHead(int majorType, int argument) {
  if (argument < 0) throw ProtocolCodecError('CBOR head argument must be non-negative');
  final mt = majorType << 5;
  if (argument < 24) return Uint8List.fromList([mt | argument]);
  if (argument <= 0xff) return Uint8List.fromList([mt | 24, argument]);
  if (argument <= 0xffff) {
    final out = Uint8List(3);
    out[0] = mt | 25;
    ByteData.sublistView(out).setUint16(1, argument, Endian.big);
    return out;
  }
  if (argument <= 0xffffffff) {
    final out = Uint8List(5);
    out[0] = mt | 26;
    ByteData.sublistView(out).setUint32(1, argument, Endian.big);
    return out;
  }
  throw ProtocolCodecError('CBOR argument exceeds 32-bit range used by this host codec');
}

Uint8List encodeUint(int value) => encodeHead(_majorUint, value);

Uint8List encodeInt(int value) =>
    value >= 0 ? encodeHead(_majorUint, value) : encodeHead(_majorNegint, -1 - value);

Uint8List encodeBytes(Uint8List value) => _concat([encodeHead(_majorBytes, value.length), value]);

Uint8List encodeText(String value) {
  final utf8Bytes = Uint8List.fromList(utf8.encode(value));
  return _concat([encodeHead(_majorText, utf8Bytes.length), utf8Bytes]);
}

Uint8List encodeBool(bool value) => Uint8List.fromList([value ? 0xf5 : 0xf4]);

Uint8List encodeFloat64(double value) {
  final out = Uint8List(9);
  out[0] = 0xfb;
  ByteData.sublistView(out).setFloat64(1, value, Endian.big);
  return out;
}

Uint8List encodeMap(List<(int, Uint8List)> pairs) {
  final parts = <Uint8List>[Uint8List.fromList([0xbf])];
  for (final pair in pairs) {
    parts.add(encodeUint(pair.$1));
    parts.add(pair.$2);
  }
  parts.add(Uint8List.fromList([0xff]));
  return _concat(parts);
}

sealed class CborValue {
  const CborValue();
}

class CborUint extends CborValue {
  const CborUint(this.value);
  final int value;
}

class CborInt extends CborValue {
  const CborInt(this.value);
  final int value;
}

class CborBytes extends CborValue {
  const CborBytes(this.value);
  final Uint8List value;
}

class CborText extends CborValue {
  const CborText(this.value);
  final String value;
}

class CborBool extends CborValue {
  const CborBool(this.value);
  final bool value;
}

class CborFloat extends CborValue {
  const CborFloat(this.value);
  final double value;
}

class CborNull extends CborValue {
  const CborNull();
}

class CborMap extends CborValue {
  const CborMap(this.value);
  final Map<int, CborValue> value;
}

class CborArray extends CborValue {
  const CborArray(this.value);
  final List<CborValue> value;
}

class CborReader {
  CborReader(this._bytes);
  final Uint8List _bytes;
  var _offset = 0;

  int get remaining => _bytes.length - _offset;

  int _readByte() {
    if (_offset >= _bytes.length) throw ProtocolCodecError('unexpected end of CBOR input');
    return _bytes[_offset++];
  }

  Uint8List _readBytes(int length) {
    if (_offset + length > _bytes.length) throw ProtocolCodecError('unexpected end of CBOR input');
    final out = Uint8List.sublistView(_bytes, _offset, _offset + length);
    _offset += length;
    return out;
  }

  int _readArgument(int additional) {
    if (additional < 24) return additional;
    if (additional == 24) return _readByte();
    if (additional == 25) {
      final b = _readBytes(2);
      return ByteData.sublistView(b).getUint16(0, Endian.big);
    }
    if (additional == 26) {
      final b = _readBytes(4);
      return ByteData.sublistView(b).getUint32(0, Endian.big);
    }
    throw ProtocolCodecError('unsupported CBOR additional info $additional');
  }

  bool get _atBreak => remaining > 0 && _bytes[_offset] == 0xff;

  CborValue readValue() {
    final head = _readByte();
    final major = head >> 5;
    final additional = head & 0x1f;
    switch (major) {
      case _majorUint:
        return CborUint(_readArgument(additional));
      case _majorNegint:
        return CborInt(-1 - _readArgument(additional));
      case _majorBytes:
        return CborBytes(_readBytes(_readArgument(additional)));
      case _majorText:
        return CborText(utf8.decode(_readBytes(_readArgument(additional))));
      case _majorArray:
        final items = <CborValue>[];
        if (additional == 31) {
          while (!_atBreak) {
            items.add(readValue());
          }
          _readByte();
        } else {
          final length = _readArgument(additional);
          for (var i = 0; i < length; i++) {
            items.add(readValue());
          }
        }
        return CborArray(items);
      case _majorMap:
        final map = <int, CborValue>{};
        void readPair() {
          final key = readValue();
          if (key is! CborUint) throw ProtocolCodecError('CBOR map key must be a non-negative integer');
          if (map.containsKey(key.value)) throw ProtocolCodecError('duplicate CBOR map key ${key.value}');
          map[key.value] = readValue();
        }

        if (additional == 31) {
          while (!_atBreak) {
            readPair();
          }
          _readByte();
        } else {
          final length = _readArgument(additional);
          for (var i = 0; i < length; i++) {
            readPair();
          }
        }
        return CborMap(map);
      case _majorSimple:
        if (additional == 20) return const CborBool(false);
        if (additional == 21) return const CborBool(true);
        if (additional == 22) return const CborNull();
        if (additional == 27) {
          final b = _readBytes(8);
          return CborFloat(ByteData.sublistView(b).getFloat64(0, Endian.big));
        }
        throw ProtocolCodecError('unsupported CBOR simple value $additional');
      default:
        throw ProtocolCodecError('unsupported CBOR major type $major');
    }
  }
}

CborValue decodeOne(Uint8List bytes) => CborReader(bytes).readValue();

int requireUint(Map<int, CborValue> map, int key, String ctx) {
  final value = map[key];
  if (value is! CborUint) throw ProtocolCodecError('$ctx key $key must be a uint');
  return value.value;
}

String requireText(Map<int, CborValue> map, int key, String ctx) {
  final value = map[key];
  if (value is! CborText) throw ProtocolCodecError('$ctx key $key must be text');
  return value.value;
}

Uint8List requireBytes(Map<int, CborValue> map, int key, String ctx) {
  final value = map[key];
  if (value is! CborBytes) throw ProtocolCodecError('$ctx key $key must be bytes');
  return value.value;
}

Object? cborToDart(CborValue value) {
  return switch (value) {
    CborUint(:final value) => value,
    CborInt(:final value) => value,
    CborBool(:final value) => value,
    CborFloat(:final value) => value,
    CborText(:final value) => value,
    CborNull() => null,
    CborBytes(:final value) => value,
    CborMap(:final value) => {for (final e in value.entries) e.key: cborToDart(e.value)},
    CborArray(:final value) => [for (final item in value) cborToDart(item)],
  };
}

Uint8List encodeDart(Object? value) {
  if (value == null) return Uint8List.fromList([0xf6]);
  if (value is bool) return encodeBool(value);
  if (value is int) return encodeInt(value);
  if (value is double) return encodeFloat64(value);
  if (value is String) return encodeText(value);
  throw ProtocolCodecError('unsupported field value ${value.runtimeType}');
}

File: software/micro-flow-editor/packages/protocol-sdk/src/cbor.ts
/**
 * Minimal canonical CBOR primitives — encode/decode only the subset the
 * firmware's zcbor-based Protocol V1 actually uses: unsigned/negative
 * integers (major type 0/1, up to 64 bits), byte strings, UTF-8 text
 * strings, definite-length arrays/maps, and the two boolean simple values.
 * No floats, no tags, no indefinite-length items, no bignums — the firmware
 * never emits them (see software/reference or the S021 implementation note
 * for the research this is based on), so a general-purpose CBOR library
 * would accept/produce a strictly larger surface than this protocol allows.
 * Encoding is always canonical (RFC 8949 §4.2.1: shortest-possible integer
 * head, definite lengths, ascending small-integer map keys) to match
 * zcbor's canonical output byte-for-byte.
 */

export class ProtocolCodecError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "ProtocolCodecError";
  }
}

const MAJOR_UINT = 0;
const MAJOR_NEGINT = 1;
const MAJOR_BYTES = 2;
const MAJOR_TEXT = 3;
const MAJOR_ARRAY = 4;
const MAJOR_MAP = 5;
const MAJOR_SIMPLE = 7;

function concatBytes(chunks: readonly Uint8Array[]): Uint8Array {
  const total = chunks.reduce((n, c) => n + c.length, 0);
  const out = new Uint8Array(total);
  let offset = 0;
  for (const chunk of chunks) {
    out.set(chunk, offset);
    offset += chunk.length;
  }
  return out;
}

/** Canonical minimal-length head (major type + argument) per RFC 8949 §3.1/§4.2.1. */
function encodeHead(majorType: number, argument: bigint): Uint8Array {
  if (argument < 0n) {
    throw new ProtocolCodecError("CBOR head argument must be non-negative");
  }
  const mt = majorType << 5;
  if (argument < 24n) {
    return Uint8Array.of(mt | Number(argument));
  }
  if (argument <= 0xffn) {
    return Uint8Array.of(mt | 24, Number(argument));
  }
  if (argument <= 0xffffn) {
    const bytes = new Uint8Array(3);
    bytes[0] = mt | 25;
    new DataView(bytes.buffer).setUint16(1, Number(argument), false);
    return bytes;
  }
  if (argument <= 0xffffffffn) {
    const bytes = new Uint8Array(5);
    bytes[0] = mt | 26;
    new DataView(bytes.buffer).setUint32(1, Number(argument), false);
    return bytes;
  }
  if (argument <= 0xffffffffffffffffn) {
    const bytes = new Uint8Array(9);
    bytes[0] = mt | 27;
    new DataView(bytes.buffer).setBigUint64(1, argument, false);
    return bytes;
  }
  throw new ProtocolCodecError("CBOR argument exceeds the 64-bit range this protocol supports");
}

export function encodeUint(value: bigint): Uint8Array {
  return encodeHead(MAJOR_UINT, value);
}

/** Encodes a signed integer of either sign as canonical CBOR major type 0/1. */
export function encodeInt(value: bigint): Uint8Array {
  return value >= 0n ? encodeHead(MAJOR_UINT, value) : encodeHead(MAJOR_NEGINT, -1n - value);
}

export function encodeBytes(value: Uint8Array): Uint8Array {
  return concatBytes([encodeHead(MAJOR_BYTES, BigInt(value.length)), value]);
}

export function encodeText(value: string): Uint8Array {
  const utf8 = new TextEncoder().encode(value);
  return concatBytes([encodeHead(MAJOR_TEXT, BigInt(utf8.length)), utf8]);
}

/**
 * Indefinite-length array start (major type 4, additional info 31) —
 * verified against the actual firmware build (not just its source): zcbor
 * in this build does **not** use canonical definite-length collections, it
 * emits `0x9F <items...> 0xFF` for arrays and `0xBF <pairs...> 0xFF` for
 * maps. This was confirmed by building and running
 * `firmware/core/tests/protocol` in `native_sim` and inspecting the actual
 * encoded bytes — the initial source-reading-only research had assumed
 * canonical/definite-length, which turned out to be wrong.
 */
export function encodeArrayHeader(): Uint8Array {
  return Uint8Array.of((MAJOR_ARRAY << 5) | 31);
}

export function encodeMapHeader(): Uint8Array {
  return Uint8Array.of((MAJOR_MAP << 5) | 31);
}

export function encodeBreak(): Uint8Array {
  return Uint8Array.of((MAJOR_SIMPLE << 5) | 31);
}

export function encodeBool(value: boolean): Uint8Array {
  return Uint8Array.of(value ? 0xf5 : 0xf4);
}

/** CBOR simple value 22 (`0xF6`) — the wire `null`. Used by `config_cbor.c`'s `encode_optional_u8` for an unspecified `bay_id`/`power_rail_id`. */
export function encodeNull(): Uint8Array {
  return Uint8Array.of(0xf6);
}

/** Concatenates an ordered sequence of already-encoded CBOR items into one buffer. */
export function encodeSequence(...items: readonly Uint8Array[]): Uint8Array {
  return concatBytes(items);
}

/**
 * Builds an indefinite-length map from `[key, valueBytes]` pairs, in the
 * order given (ascending field ID, matching the firmware's own field
 * ordering) — see `encodeMapHeader`'s doc comment for why this is
 * indefinite-length rather than canonical.
 */
export function encodeMap(pairs: ReadonlyArray<readonly [number, Uint8Array]>): Uint8Array {
  const parts: Uint8Array[] = [encodeMapHeader()];
  for (const [key, value] of pairs) {
    parts.push(encodeUint(BigInt(key)), value);
  }
  parts.push(encodeBreak());
  return concatBytes(parts);
}

export function encodeArray(items: readonly Uint8Array[]): Uint8Array {
  return concatBytes([encodeArrayHeader(), ...items, encodeBreak()]);
}

export type CborValue =
  | { readonly kind: "uint"; readonly value: bigint }
  | { readonly kind: "int"; readonly value: bigint }
  | { readonly kind: "bytes"; readonly value: Uint8Array }
  | { readonly kind: "text"; readonly value: string }
  | { readonly kind: "array"; readonly value: readonly CborValue[] }
  | { readonly kind: "map"; readonly value: ReadonlyMap<number, CborValue> }
  | { readonly kind: "bool"; readonly value: boolean }
  | { readonly kind: "null" };

/** Cursor-based decoder for one CBOR value at a time, tracking consumed bytes. */
export class CborReader {
  private offset = 0;

  constructor(private readonly bytes: Uint8Array) {}

  get remaining(): number {
    return this.bytes.length - this.offset;
  }

  private readByte(): number {
    if (this.offset >= this.bytes.length) {
      throw new ProtocolCodecError("unexpected end of CBOR input");
    }
    return this.bytes[this.offset++]!;
  }

  private readBytes(length: number): Uint8Array {
    if (this.offset + length > this.bytes.length) {
      throw new ProtocolCodecError("unexpected end of CBOR input");
    }
    const out = this.bytes.subarray(this.offset, this.offset + length);
    this.offset += length;
    return out;
  }

  private readArgument(additionalInfo: number): bigint {
    if (additionalInfo < 24) return BigInt(additionalInfo);
    if (additionalInfo === 24) return BigInt(this.readByte());
    if (additionalInfo === 25) {
      const b = this.readBytes(2);
      return BigInt(new DataView(b.buffer, b.byteOffset, 2).getUint16(0, false));
    }
    if (additionalInfo === 26) {
      const b = this.readBytes(4);
      return BigInt(new DataView(b.buffer, b.byteOffset, 4).getUint32(0, false));
    }
    if (additionalInfo === 27) {
      const b = this.readBytes(8);
      return new DataView(b.buffer, b.byteOffset, 8).getBigUint64(0, false);
    }
    throw new ProtocolCodecError(`unsupported CBOR additional info ${additionalInfo}`);
  }

  /** True if the next byte is the indefinite-length "break" marker (0xFF) — does not consume it. */
  private atBreak(): boolean {
    return this.remaining > 0 && this.bytes[this.offset] === 0xff;
  }

  readValue(): CborValue {
    const head = this.readByte();
    const majorType = head >> 5;
    const additionalInfo = head & 0x1f;
    switch (majorType) {
      case MAJOR_UINT:
        return { kind: "uint", value: this.readArgument(additionalInfo) };
      case MAJOR_NEGINT:
        return { kind: "int", value: -1n - this.readArgument(additionalInfo) };
      case MAJOR_BYTES:
        return { kind: "bytes", value: this.readBytes(Number(this.readArgument(additionalInfo))) };
      case MAJOR_TEXT: {
        const raw = this.readBytes(Number(this.readArgument(additionalInfo)));
        return { kind: "text", value: new TextDecoder("utf-8", { fatal: true }).decode(raw) };
      }
      case MAJOR_ARRAY: {
        const items: CborValue[] = [];
        if (additionalInfo === 31) {
          // Indefinite-length — the shape the actual firmware build emits
          // (see `encodeArrayHeader`'s doc comment); read until the break.
          while (!this.atBreak()) items.push(this.readValue());
          this.readByte(); // consume 0xFF
        } else {
          const length = Number(this.readArgument(additionalInfo));
          for (let i = 0; i < length; i++) items.push(this.readValue());
        }
        return { kind: "array", value: items };
      }
      case MAJOR_MAP: {
        const map = new Map<number, CborValue>();
        const readPair = () => {
          const key = this.readValue();
          if (key.kind !== "uint") {
            throw new ProtocolCodecError("CBOR map key must be a non-negative integer");
          }
          const numericKey = Number(key.value);
          if (map.has(numericKey)) {
            throw new ProtocolCodecError(`duplicate CBOR map key ${numericKey}`);
          }
          map.set(numericKey, this.readValue());
        };
        if (additionalInfo === 31) {
          while (!this.atBreak()) readPair();
          this.readByte(); // consume 0xFF
        } else {
          const length = Number(this.readArgument(additionalInfo));
          for (let i = 0; i < length; i++) readPair();
        }
        return { kind: "map", value: map };
      }
      case MAJOR_SIMPLE:
        if (additionalInfo === 20) return { kind: "bool", value: false };
        if (additionalInfo === 21) return { kind: "bool", value: true };
        if (additionalInfo === 22) return { kind: "null" };
        throw new ProtocolCodecError(`unsupported CBOR simple value ${additionalInfo}`);
      default:
        throw new ProtocolCodecError(`unsupported CBOR major type ${majorType}`);
    }
  }
}

export function decodeOne(bytes: Uint8Array): CborValue {
  return new CborReader(bytes).readValue();
}
